### ISSUE TITLE
Sly completion error

### ISSUE THREAD

**Issue by mfiano** [joaotavora/sly#226]
Sly completion is broken when trying to use `company-sly` to complete *un-exported* symbols if the Spinneret system is loaded into an image. This is a very isolated case, and I'm unsure which project to report the bug to. I started here, because there doesn't seem to be any suspicious package binding activity in Spinneret.

Recipe:
1) M-x sly
2) In REPL, issue `(ql:quickload :spinneret)`.
3) In REPL, type `spinneret::` followed by a TAB or whatever is bound to company for completion. Also note the two trailing colons for un-exported symbols. Also note that this errors when completing any package with a double colon suffix, not just spinneret. The result is the following backtrace:

```
The name NIL does not designate any package.
   [Condition of type PACKAGE-DOES-NOT-EXIST]

Restarts:
 0: [CONTINUE] Use the current package, COMMON-LISP-USER.
 1: [RETRY] Retry finding the package.
 2: [USE-VALUE] Specify a different package
 3: [*ABORT] Return to SLY's top level.
 4: [ABORT] abort thread (#<THREAD "slynk-worker" RUNNING {1002DE66A3}>)

Backtrace:
 0: (SB-INT:%FIND-PACKAGE-OR-LOSE NIL)
 1: (SB-INT:FIND-UNDELETED-PACKAGE-OR-LOSE NIL)
 2: (PACKAGE-NICKNAMES NIL)
 3: ((FLET SLYNK-COMPLETION::SORTED-NICKNAMES :IN SLYNK-COMPLETION::QUALIFIED-MATCHING) NIL)
 4: (SLYNK-COMPLETION::QUALIFIED-MATCHING "SPINNERET::" #<PACKAGE "COMMON-LISP-USER">)
 5: (SLYNK-COMPLETION:FLEX-COMPLETIONS "spinneret::" "common-lisp-user" :LIMIT 300)
 6: (SB-INT:SIMPLE-EVAL-IN-LEXENV (SLYNK-COMPLETION:FLEX-COMPLETIONS "spinneret::" (QUOTE "common-lisp-user")) #<NULL-LEXENV>)
 7: (EVAL (SLYNK-COMPLETION:FLEX-COMPLETIONS "spinneret::" (QUOTE "common-lisp-user")))
 8: (SLYNK:EVAL-FOR-EMACS (SLYNK-COMPLETION:FLEX-COMPLETIONS "spinneret::" (QUOTE "common-lisp-user")) "common-lisp-user" 40)
 9: ((LAMBDA NIL :IN SLYNK::SPAWN-WORKER-THREAD))
10: (SLYNK-SBCL::CALL-WITH-BREAK-HOOK #<FUNCTION SLYNK:SLYNK-DEBUGGER-HOOK> #<FUNCTION (LAMBDA NIL :IN SLYNK::SPAWN-WORKER-THREAD) {52D3CE4B}>)
11: ((FLET SLYNK-BACKEND:CALL-WITH-DEBUGGER-HOOK :IN "/home/mfiano/.emacs.d/straight/repos/sly/slynk/backend/sbcl.lisp") #<FUNCTION SLYNK:SLYNK-DEBUGGER-HOOK> #<FUNCTION (LAMBDA NIL :IN SLYNK::SPAWN-WORKE..
12: ((LAMBDA NIL :IN SLYNK::CALL-WITH-LISTENER))
13: (SLYNK::CALL-WITH-BINDINGS ((*PACKAGE* . #<PACKAGE "COMMON-LISP-USER">) (* . #1=(:SPINNERET)) (**) (***) (/ #1#) (//) ...) #<CLOSURE (LAMBDA NIL :IN SLYNK::CALL-WITH-LISTENER) {1002DF818B}>)
14: ((LAMBDA NIL :IN SLYNK::SPAWN-WORKER-THREAD))
15: ((FLET SB-UNIX::BODY :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
16: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
17: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
18: ((FLET "WITHOUT-INTERRUPTS-BODY-1" :IN SB-THREAD::CALL-WITH-MUTEX))
19: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {147B6D2A6D7B}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THR..
20: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "slynk-worker" RUNNING {1002DE66A3}> NIL #<CLOSURE (LAMBDA NIL :IN SLYNK::SPAWN-WORKER-THREAD) {1002DE664B}> NIL)
21: ("foreign function: call_into_lisp")
22: ("foreign function: new_thread_trampoline")
```

**Comment by mfiano** [comment]
Further investigation shows that slynk assumes `do-all-symbols` only operates on symbols that have a home package. If we look at the locals of frame 4 above, we see the problem:

```lisp
 4: (SLYNK-COMPLETION::QUALIFIED-MATCHING "SPINNERET::" #<PACKAGE "COMMON-LISP-USER">)
      Locals:
        COLLECTED = #<HASH-TABLE :TEST EQL :COUNT 378 {1003EC0003}>
        HOME-PACKAGE = #<PACKAGE "COMMON-LISP-USER">
        NICKNAMES-BY-PACKAGE = #<HASH-TABLE :TEST EQL :COUNT 117 {1003E97B53}>
        PATTERN = "SPINNERET::"
        RET = #<HASH-TABLE :TEST EQL :COUNT 0 {1003E97733}>
        S#1 = #:OPEN
```

`#:OPEN` doesn't have a package.

**Comment by mfiano** [comment]
wrapping the body of the `sorted-nicknames` local function with `(when package ...)` fixes it.

**Comment by joaotavora** [comment]
I was going to do the fix you suggested but then I noticed this in

http://clhs.lisp.se/Body/m_do_sym.htm

"do-all-symbols iterates on every registered package. do-all-symbols will not process every symbol whatsoever, because a symbol not accessible in any registered package will not be processed. do-all-symbols may cause a symbol that is present in several packages to be processed more than once. "

Apparently, in sbcl, "accessible in a registered package" includes symbols that don't have a package. I don't know how that can be, how can your uninterned access your uninterned `#:open` be accessed??  if I type `#:open` in the REPL it's a different `#:open`.

In other words

```
CL-USER> (eq '#:open '#:open)
NIL
```

Maybe I'm missing something obvious, or some CL pope can enlighten.

Anyway, your fix seems suitable.

**Comment by joaotavora** [comment]
Thanks, btw!

**Comment by phoe** [comment]
\*I put on my robe and Lisp pope hat\*

> Apparently, in sbcl, "accessible in a registered package" includes symbols that don't have a package. I don't know how that can be, how can your uninterned access your uninterned `#:open` be accessed??

This behaviour is simple and actually standard and not SBCL-specific. You create a symbol `A::FOO`, import it into package `B`, and then unintern it from package `A`. The symbol named `FOO` is therefore accessible in package `B` even though it has no home package.

See http://www.flownet.com/gat/packages.pdf for another example of that:

![Zrzut ekranu z 2019-05-05 23-09-50](https://user-images.githubusercontent.com/15045546/57200329-f2221280-6f8a-11e9-842b-933d913123d6.png)

As for why this manifests on SBCL in particular and not e.g. on CCL, I am still investigating. It is likely because of the implicit package use lists and how Parenscript specially treats `CL` symbols in https://gitlab.common-lisp.net/parenscript/parenscript/blob/master/src/js-dom-symbol-exports.lisp#L49 - for instance, if a `:USE` clause in a `DEFPACKAGE` form is omitted, CCL puts packages `CCL` and `CL` in the package use list implicitly; SBCL seems to have an implicit default use list. And this seems to interact with shadowing-importing.

**Comment by joaotavora** [comment]
> This behaviour is simple and actually standard and not SBCL-specific. You create a symbol A::FOO, import it into package B, and then unintern it from package A. The symbol named FOO is therefore accessible in package B even though it has no home package.

> *I put on my robe and Lisp pope hat*

Perfectly clear and thanks for the explanation! :pray:

**Comment by phoe** [comment]
No problem!

\*I take off my robe and Lisp pope hat\*
